lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -299,9 +348,150 @@ class YieldOnceCheck : public SILFunctionTransform {
+           llvm::breadth_first<llvm::Inverse<SILBasicBlock *>>(noYieldPred)) {
+        if (!visitedBBs.count(pred)) {
+          continue;
+        }
+        if (predcessorsOfYieldPred.count(pred)) {
+          lowestCommonAncestorBB = pred;
+          break;
+        }
+        predcessorsOfNoYieldPred.insert(pred);
+      }
+      assert(lowestCommonAncestorBB);
+
+      auto *conflictingBranch = lowestCommonAncestorBB->getTerminator();
+
+      // Step 2: Find the target basic block of the 'conflictingBranch' that
+      // doesn't yield.
+      SILBasicBlock *noYieldTarget = nullptr;
+      for (auto *succ : lowestCommonAncestorBB->getSuccessorBlocks()) {
+        if (predcessorsOfNoYieldPred.count(succ)) {
+          noYieldTarget = succ;
+          break;
+        }
+      }
+      assert(noYieldTarget);
+
+      // Step 3: Report specialized diagnostics for each kind of conflicting
+      // branch.
+
+      // For conditional-branch instructions, which correspond to the
+      // conditions of 'if', 'where' or 'guard' statements, report for what
+      // truth value the branch doesn't yield.
+      if (auto *condbr = dyn_cast<CondBranchInst>(conflictingBranch)) {
+        auto diag = (condbr->getTrueBB() == noYieldTarget)
+                        ? diag::true_branch_doesnt_yield
+                        : diag::false_branch_doesnt_yield;
+        diagnose(astCtx, condbr->getLoc().getSourceLoc(), diag);
+        return;
+      }
+
+      // For switch_enum instructions, report the case that doesn't yield.
+      if (auto *switchEnum = dyn_cast<SwitchEnumInstBase>(conflictingBranch)) {
+        if (switchEnum->hasDefault() &&
+            switchEnum->getDefaultBB() == noYieldTarget) {
+          diagnose(astCtx, noYieldTarget->begin()->getLoc().getSourceLoc(),
+                   diag::default_case_doesnt_yield);
+          return;
+        }
+        // Find the case that doesn't yield.
+        auto enumElemDecl =
+            switchEnum->getUniqueCaseForDestination(noYieldTarget);
+        assert(enumElemDecl.isNonNull());
+
+        diagnose(astCtx, noYieldTarget->begin()->getLoc().getSourceLoc(),
+                 diag::switch_case_doesnt_yield,
+                 enumElemDecl.get()->getNameStr());
+        return;
+      }
+
+      // For switch_value instructions, report the case number that doesn't
+      // yield.
+      if (auto *switchValue = dyn_cast<SwitchValueInst>(conflictingBranch)) {
+        if (switchValue->hasDefault() &&
+            switchValue->getDefaultBB() == noYieldTarget) {
+          diagnose(astCtx, noYieldTarget->begin()->getLoc().getSourceLoc(),
+                   diag::default_case_doesnt_yield);
+          return;
+        }
+        // Find the case that doesn't yield.
+        auto caseNumber =
+            switchValue->getUniqueCaseForDestination(noYieldTarget);
+        assert(caseNumber.hasValue());
+
+        diagnose(astCtx, noYieldTarget->begin()->getLoc().getSourceLoc(),
```

Comment:
Instead of using the raw case number here ("in case 2") you should use llvm::StringExtra.h's getOrdinalSuffix() to say "in the 2nd case". This reads more naturally and also helps reduce confusion about wether the index is 0-based or 1-based.